Ruma mono

DevinR528 opened the pull request:
An example of moving to ruma-events 0.22 and the ruma monorepo.

vbrandl committed:
Implement display name resolving

vbrandl committed:
Rename accessor for display name

vbrandl committed:
Fix failing test

vbrandl committed:
Only add name duplicates to the display name map

vbrandl committed:
Add first test for display names

dkasak committed:
Rename display_names -> disambiguated_display_names.

dkasak committed:
Fix Markdown in doc comment.

dkasak committed:
matrix-sdk: Vary sync token with each EventBuilder::build_sync_response call.

This allows us to hold onto an EventBuilder object and use it to build
multiple sync responses. Previously this would have resulted in each
of the responses having the same next_batch sync token. This would make
clients ignore the latter responses if they have already received any of
the previous ones.

dkasak committed:
Properly test for display name disambiguation.

dkasak committed:
Fix display name disambiguation so it passes the test.

dkasak committed:
Use "disambiguated" instead of "resolved" display name in the doc comment.

To match how the C2S spec calls it.

dkasak committed:
Add short explanation to Room::member_display_name.

dkasak committed:
Make Room::member_display_name return MXID as fallback.

If there is no display name set. This means the method can now always
return something so there is no need to wrap in an `Option`.

dkasak committed:
Run cargo fmt.

dkasak committed:
Add explicit type annotation.

dkasak committed:
Fix comment styling.

dkasak committed:
fixup: explicit type annotations

dkasak committed:
Retrieve user id from RoomMember instead of reconstructing.

dkasak committed:
Refactor member_display_name.

Make it more readable, add comments.

dkasak committed:
Add doc comment to RoomName::calculate_name.

dkasak committed:
Small refactor so we don't duplicate user_id creation.

dkasak committed:
Style fixes.

dkasak committed:
Add RoomMember::name.

Returns the most ergonomic name for the member (either the display name
(if set) or the MXID).

dkasak committed:
Implement RoomMember::unique_name.

This gives us a name that is as ergonomic as possible while guaranteeing
it is unique.

dkasak committed:
Nix RoomMember::update_member and tracking membership.

After discussing with poljar, we concluded we don't actually need to
tracking membership state, since we won't be tracking users that
left (banned, kicked, disinvited).

The only thing we need to keep track of is the difference between joined
and invited users which will be dealt with in a separate commit.

dkasak committed:
Split joined/invited users and handle removing users.

dkasak committed:
Must not take our user into account when calculating room name.

DevinR528 commented on matrix_sdk_crypto/src/machine.rs:
```diff
@@ -930,10 +928,30 @@ impl OlmMachine {
+            "sender": sender
+        });
+
+        let event = serde_json::from_value::<EventJson<AnyToDeviceEvent>>(device_event)?;
```
The old `ToDevice` event had a custom `Deserialize` impl so picking apart the object is needed now.

poljar replied:
Aren't you here pulling out fields that are inside a json value and then constructing the json value with the same format and data? How do the two json values differ?

DevinR528 replied:
The derive `Deserialize` will fail if it finds a key that isn't in the rust struct (the 0.22 ToDeviceEvent struct derives `Deserialize`). This failed to deserialize because it has other fields. It took me a while to realize that the custom one https://github.com/ruma/ruma-events/blob/0.21.2/src/to_device.rs#L148 was just picking appart and ignoring the rest.

DevinR528 commented on matrix_sdk_crypto/src/machine.rs:
```diff
@@ -1011,41 +1030,38 @@ impl OlmMachine {
+                        sender: sender.clone(),
+                    };
+
+                    Ok(Some(EventJson::from(event)))
```
Does this fix the TODO?

poljar replied:
If the tests pass I would say yes.

DevinR528 commented on matrix_sdk_crypto/src/machine.rs:
```diff
@@ -932,10 +928,30 @@ impl OlmMachine {
-            serde_json::from_value::<EventJson<ToDeviceEvent>>(decrypted_json)?,
-            signing_key.to_owned(),
-        ))
+        let ev_type = decrypted_json
+            .get("type")
+            .cloned()
+            .ok_or_else(|| EventError::MissingField("type".to_string()))?;
+
+        let content = decrypted_json
+            .get("content")
+            .cloned()
+            .ok_or_else(|| EventError::MissingField("content".to_string()))?;
+
+        let sender = decrypted_json
+            .get("sender")
+            .cloned()
+            .ok_or_else(|| EventError::MissingField("sender".to_string()))?;
+
+        let device_event = json!({
+            "type": ev_type,
+            "content": content,
+            "sender": sender
+        });
+
+        let event = serde_json::from_value::<EventJson<AnyToDeviceEvent>>(device_event)?;
+
+        Ok((event, signing_key.to_owned()))
```
The old `ToDevice` had a custom `Deserialize` impl so now that it is just derived it has to be picked apart.

DevinR528 commented on matrix_sdk_crypto/src/machine.rs:
```diff
@@ -1010,38 +1030,38 @@ impl OlmMachine {
-                // let event = EventJson::from(ToDeviceEvent::RoomKey(event.clone()));
-                // This saidly lacks a type once it's serialized again, fix
-                // this in Ruma.
-                let mut json = serde_json::to_value(event.clone())?;
-                json.as_object_mut()
-                    .unwrap()
-                    .insert("type".to_owned(), Value::String("m.room_key".to_owned()));
-                let event = serde_json::from_value::<EventJson<ToDeviceEvent>>(json)?;
-
-                Ok(Some(event))
-            }
+        content: &mut AnyToDeviceEventContent,
+    ) -> OlmResult<Option<EventJson<AnyToDeviceEvent>>> {
+        match content {
+            AnyToDeviceEventContent::RoomKey(c) => match &mut c.algorithm {
+                Algorithm::MegolmV1AesSha2 => {
+                    let session_key = GroupSessionKey(mem::take(&mut c.session_key));
+
+                    let session =
+                        InboundGroupSession::new(sender_key, signing_key, &c.room_id, session_key)?;
+                    let _ = self.store.save_inbound_group_session(session).await?;
+
+                    let event = AnyToDeviceEvent {
+                        content: content.clone(),
+                        sender: sender.clone(),
+                    };
```
Does this resolve the TODO

DevinR528 commented on test_data/leave_sync.json:
```diff
@@ -188,7 +187,6 @@
                                 "membership": "join"
                             },
                             "event_id": "$152034824468gOeNB:localhost",
-                            "membership": "join",
```
Not sure how this got in or if this is ok according to the spec but it broke ruma's ser/de

poljar replied:
I think this might be some old event format, while it's fine to ignore it I don't think Ruma should break if it finds such a field.

DevinR528 replied:
I'm surprised `serde` fails to deserialize when it finds extra keys. Before ruma-events had optional room_id fileds this might make more sense?

DevinR528 replied:
I agree, there isn't much security risk in ignoring fields. I wonder if serde has an option to not fail unless a field that should be there isn't?

poljar replied:
Ignoring them is required for forwards compatibility. Serde should already do that by default: https://serde.rs/container-attrs.html#deny_unknown_fields

jplatte replied:
This is probably because we don't use `serde`s `Deserialize` derive and missed this case when generating `Deserialize` implementations in ruma-events-macros. I would suspect a `#[serde(other)]` variant for our [`Field` enum](https://github.com/ruma/ruma/blob/3ec0936a2818314f490b9867eb3040c56719f2ed/ruma-events-macros/src/event.rs#L215-L221) fixes this.

DevinR528 commented on test_data/sync.json:
```diff
@@ -200,31 +197,7 @@
-                              "prev_sender": "@example2:localhost",
-                              "age": 6992
-                            },
-                            "room_id": "!roomid:room.com"
```
There are no room_id's in sync responses right?

poljar replied:
They are not required per spec and Synapse doesn't include them, now other servers might include them

DevinR528 replied:
Yeah it's not great that serde fails when it encounters a unknown key in the JSON. In the ruma-events matrix-sdk actually uses room_id is `Option<RoomId>` maybe that's the way to do it?

DevinR528 commented on matrix_sdk_crypto/src/machine.rs:
```diff
@@ -932,10 +928,30 @@ impl OlmMachine {
-            signing_key.to_owned(),
-        ))
+        let ev_type = decrypted_json
+            .get("type")
+            .cloned()
+            .ok_or_else(|| EventError::MissingField("type".to_string()))?;
+
+        let content = decrypted_json
+            .get("content")
+            .cloned()
+            .ok_or_else(|| EventError::MissingField("content".to_string()))?;
+
+        let sender = decrypted_json
+            .get("sender")
+            .cloned()
+            .ok_or_else(|| EventError::MissingField("sender".to_string()))?;
+
+        let device_event = json!({
+            "type": ev_type,
+            "content": content,
+            "sender": sender
+        });
+
+        let event = serde_json::from_value::<EventJson<AnyToDeviceEvent>>(device_event)?;
```
ToDevice custom deser

DevinR528 commented on matrix_sdk_base/src/client.rs:
```diff
@@ -1406,41 +1612,155 @@ impl BaseClient {
-                event_emitter
-                    .on_unrecognized_event(room, &CustomOrRawEvent::CustomRoom(custom))
-                    .await
+            AnyRoomEventStub::StateEvent(event) => match &event.content {
+                AnyStateEventContent::RoomMember(mem) => {
+                    let prev = event
+                        .prev_content
+                        .as_ref()
+                        .map(|p| {
+                            if let AnyStateEventContent::RoomMember(m) = p {
+                                Some(m.clone())
+                            } else {
+                                None
+                            }
+                        })
+                        .flatten();
+                    event_emitter
+                        .on_room_member(room, &map_state_stub(event, mem.clone(), prev))
+                        .await
+                }
```
This allows the EventEmitter to receive events with a specific content field as apposed to `StateEvent<AnyStateEventContent>`. It does add a lot of code but there was talk of adding a `map_content` method to the event structs.

poljar replied:
I'm not 100% sure if we want this, I'm leaning towards yes, but this surely doesn't need to be in the same PR.

DevinR528 commented on matrix_sdk_base/src/models/room.rs:
```diff
@@ -516,20 +679,9 @@ impl Room {
     /// # Arguments
     ///
     /// * `event` - The event of the room.
-    pub fn receive_state_event(&mut self, event: &StateEvent) -> bool {
-        match event {
-            // update to the current members of the room
-            StateEvent::RoomMember(member) => self.handle_membership(member),
-            // finds all events related to the name of the room for later use
-            StateEvent::RoomName(name) => self.handle_room_name(name),
-            StateEvent::RoomCanonicalAlias(c_alias) => self.handle_canonical(c_alias),
-            StateEvent::RoomAliases(alias) => self.handle_room_aliases(alias),
-            // power levels of the room members
-            StateEvent::RoomPowerLevels(power) => self.handle_power_level(power),
-            StateEvent::RoomTombstone(tomb) => self.handle_tombstone(tomb),
-            StateEvent::RoomEncryption(encrypt) => self.handle_encryption_event(encrypt),
-            _ => false,
-        }
+    pub fn receive_state_event(&mut self, event: &AnyStateEventStub, room_id: &RoomId) -> bool {
+        // forward to the room event handlers.
+        self.receive_timeline_event(&AnyRoomEventStub::StateEvent(event.clone()), room_id)
```
It seems to me that because the types from ruma are more accurate now we could remove some of these methods and just have a `receive_sync_event`?

poljar replied:
Not sure exactly what you mean, what would the method do and what type signature would it have?

In any way, let's not add more changes to this PR.

DevinR528 replied:
Since we have the `room_id` with the sync response we could convert all the `receive_*_event` and `on_*_event` methods to `receive_*_event(&self, event: AnyRoomEvent)` by converting the `AnyRoomEventStub` to a `AnyRoomEvent`. It may remove the need for conversions.

poljar commented on matrix_sdk_base/src/client.rs:
```diff
@@ -64,6 +71,77 @@ use matrix_sdk_crypto::{CryptoStore, OlmError, OlmMachine, OneTimeKeys};
 
 pub type Token = String;
 
+pub fn map_state_stub<C: StateEventContent>(
```
Could you explain what are these used for? Why are they needed now and what does stub mean here?

DevinR528 replied:
The stub means no `room_id`. They are needed to go from `StateEventStub<AnyStateEventContent>` -> `StateEventStub<AliasesEventContent>` or put more generally `EventStruct<ContentEnum>` -> `EventStruct<SpecificEvent>`.

DevinR528 replied:
Jonas did say he wanted to figure out a way to add a `map_content` method but wanted a non boilerplate way to do it. That would remove the need for the functions.

poljar replied:
I'm a bit surprised that we want to almost double the API surface of Ruma to have separate event types that have a `room_id` field and one that don't.

Not sure if there's a way to handle this better than to either double the API surface or do those mass conversions.

poljar commented on matrix_sdk_base/src/client.rs:
```diff
@@ -673,8 +756,8 @@ impl BaseClient {
         room_id: &RoomId,
-        event: &mut EventJson<RoomEvent>,
-    ) -> Result<(Option<EventJson<RoomEvent>>, bool)> {
+        event: &mut EventJson<AnyRoomEventStub>,
```
Why are we now receiving one type but returning another type?

DevinR528 replied:
The crypto stuff needs the room_id, or at least from my skimming and not fully understanding, it seemed to need a room_id. The `OlmMachine::decrypt_room_event` then returns an event with a room_id so I convert it back to a stub so the `BaseClient`, `Room` and `EventEmitter` methods are happy.

That does seem to add a lot of complexity just to avoid `Option<RoomId>`, at least for the clients. I'm sure there are improvements to what I've got here so maybe there is a more straightforward way of handing the events around that I'm not seeing.

If he wants to stick with the `Any*EventStub` for no `room_id` and `Any*Event` for events, we could just convert every event received in the sync response to having a `room_id` then no switching back and forth from stub to normal?

poljar replied:
Well the decryption method could also take the room id as the argument. 

Though we do want to decrypt events that we get from the `/room/$roomId/messages` endpoint and since those are separate types we'll need two methods. Having two separate decryption methods doesn't sound nice either.

poljar commented on matrix_sdk/src/request_builder.rs:
```diff
@@ -83,9 +83,15 @@ impl RoomBuilder {
+    pub fn creation_content(
+        &mut self,
+        federate: bool,
+        predecessor: Option<PreviousRoom>,
```
I might be wrong since it's hard to see from the github UI but I think this is missing docs for the predecessor. 

Something to keep in mind (and certainly for another PR), we're not setting a content here so we might want to flatten this method (a separate federate() method and predecessor one) if more stuff gets added to the creation content this method will need to introduce breaking changes.

poljar commented on matrix_sdk_base/src/client.rs:
```diff
@@ -684,40 +767,68 @@ impl BaseClient {
+
+                            if let Some(o) = &mut *olm {
+                                decrypted_event = o
+                                    .decrypt_room_event(&map_message_stub(
```
This is a bit weird, we are required to convert the event to another type before we can decrypt it. Remember that the crypto crate needs to be usable on its own.

poljar commented on matrix_sdk_base/src/client.rs:
```diff
@@ -992,7 +1133,68 @@ impl BaseClient {
                 if let Some(e) = decrypted_event {
-                    *event = e;
+                    match e.deserialize() {
+                        Ok(AnyRoomEvent::State(ev)) => {
```
Why is all of this suddenly needed? State events aren't encrypted so this will never succeed, note that people might send events that have the type of `m.room.member` but this won't be a valid state event.

DevinR528 replied:
This is just to get the event back to a AnyRoomEventStub.

poljar commented on matrix_sdk_base/src/models/room.rs:
```diff
@@ -491,20 +554,120 @@ impl Room {
-            RoomEvent::RoomEncryption(encrypt) => self.handle_encryption_event(encrypt),
-            #[cfg(feature = "messages")]
-            RoomEvent::RoomMessage(msg) => self.handle_message(msg),
+            AnyRoomEventStub::StateEvent(event) => match &event.content {
```
Why are all these conversions suddenly needed?

poljar commented on matrix_sdk_crypto/src/machine.rs:
```diff
@@ -932,10 +928,30 @@ impl OlmMachine {
+            "sender": sender
+        });
+
+        let event = serde_json::from_value::<EventJson<AnyToDeviceEvent>>(device_event)?;
```
As I mentioned earlier, how does `device_event` differ from `decrypted_json` so that the de-serialization fails with one but not the other?

poljar replied:
You answered this while I was doing the review, the deserialization should not fail inside of Ruma, additional fields should always be ignored. I'm a bit confused on why this isn't so already since you need to request from serde to fail using [this](https://serde.rs/container-attrs.html#deny_unknown_fields) flag.

DevinR528 replied:
Ahhh :facepalm: dang this is a custom `Deserialize` impl I can't believe I spent so long trying to figure this out. So I guess I'll have to change that in ruma.

DevinR528 replied:
Serde's example is what the event structs Deserialize is based on and it fails when it encounters an unknown field [we used the commented out derive](https://serde.rs/deserialize-struct.html) and that is where it fails.

poljar commented on test_data/context.json:
```diff
@@ -176,7 +176,6 @@
         "membership": "join"
       },
       "event_id": "$151396611913abyeC:localhost",
-      "membership": "join",
```
Let's not remove this, it seems to be useful if deserialization is borked for additional fields that should be ignored.

DevinR528 commented:
My plan was to just show you this and talk about it then once you feel like the `ruma` monorepo is ready I'll open another and take my time converting it (also having the experience of doing it once will be nice). So I'll leave this PR open so I can refer back to it.

I need to practice my github reviewing this got crazy fast sorry about that. I guess force pushing didn't help.

poljar commented:
> My plan was to just show you this and talk about it then once you feel like the `ruma` monorepo is ready I'll open another and take my time converting it (also having the experience of doing it once will be nice). So I'll leave this PR open so I can refer back to it.

Sure, that's good. It also showcased some problems with the new Ruma event APIs. So it's a win either way.

> I need to practice my github reviewing this got crazy fast sorry about that. I guess force pushing didn't help.

No worries, it's a big change and it's hard to make it easily consumable, just make sure to keep it to one topic once the PRs get to a certain size.

DevinR528 committed:
matrix-sdk: Export matrix-sdk-base Error type as BaseError

poljar committed:
matrix-sdk: Fix the author field in the cargo files.

poljar committed:
Merge branch 'export-base-error'

dkasak committed:
Hoist prev_content to top-level in both timeline and state events.

Also refactor and document why this hoisting is needed.

This change makes the user_presence test fail because the hoisting
exposes an error encoded into the test's expected result.

Previously, the test expected 2 members in the room at the end. This is
incorrect since one of the members in the test data leaves the room.
However, since the prev_content of state events was previously not
hoisted to the top level, the `membership_change` method would not
notice it and thus not realize the member had left the room. The test
was corrected to expect only a single member in the room.

Another test change was made due to a limitation of EventBuilder: due to
the fact that it makes the test data go through a de/ser cycle, it
cannot easily hoist prev_content to the top level. Because of this, the
tests were change to put prev_content into the top level from the
outset.

dkasak committed:
Fix comment style.

dkasak committed:
Correctly handle disambiguation for exiting members, refactor and test.

DevinR528 committed:
crypto: Fix overflow in should_upload_keys, bail out if uploaded keys > max uploaded

poljar committed:
Merge branch 'upload-keys'

MTRNord committed:
Do wasm sepcific changes:
- Only use send+sync when not using wasm
- Use wasm capabale async_trait wrapper macro
- Make room and room_member specific structs always clonable

MTRNord committed:
Commit missing matrix_sdk_common_macros folder

MTRNord committed:
Fix cargo fmt issues that the local version didn't auto fix

MTRNord committed:
Add missing dependency matrix-sdk-common-macros to matrix-sdk

DevinR528 committed:
matrix-sdk-test: Remove duplicate test_data folder, leave top-level

dkasak committed:
Fix compilation error and remaining test.

Ref. for compilation error:
https://github.com/rust-lang/rust/issues/64552

dkasak committed:
Merge branch 'master' into feature/display-name

dkasak committed:
Run cargo fmt and apply clippy lints.

DevinR528 committed:
Move matrix-sdk to ruma monorepo

DevinR528 committed:
matrix_sdk_test: Use static JSON values instead of reading files

DevinR528 committed:
matrix_sdk_base: Use test_json values for tests in base

DevinR528 committed:
crypto/base: Finish moving to using static json values for test data

DevinR528 committed:
matrix_sdk: Add get_public_rooms* methods to Client

This also adds a RoomSearchBuilder for making get_public_rooms_filtered
requests and a test for each method.

poljar committed:
matrix-sdk: Remove the last test_data folder and fix the remaining tests.

poljar committed:
Merge branch 'feature/display-name'

poljar committed:
matrix-sdk: Disable the tarpaulin skip lines since it fails to run with them.

DevinR528 committed:
matrix_sdk: Rename public room builder and client methods

Remove 'get' from get_public_rooms* methods.
Rename RoomSearchBuilder -> RoomListFilterBuilder.
Use u32 over UInt in builders and Into<String> for String.
Fix docs of public room methods and builders.

DevinR528 committed:
Use ruma/ruma master, address review issues

DevinR528 committed:
Rebase upstream/master into ruma-mono branch

DevinR528 committed:
matrix-sdk-base: Use new accessor methods for models/message.rs

ruma now has field access methods for all of the Any*Event enums use
them for MessageWrapper's AnyMessageEventStub contents.

poljar committed:
crypto: Simplify the group session pair creation.

DevinR528 committed:
Move test data to test crate, fix docs

poljar committed:
Merge branch 'room-search'

poljar committed:
matrix-sdk: Remove an unused import.

poljar committed:
base: Put the decrypted event replacing in the correct place.

dkasak committed:
Add a bunch of FIXMEs to have receive_* methods do the emitting.

dkasak committed:
Fix and test hoisting of `prev_content` for timeline events.

The previous test only tested using the `EventEmitter`, which missed the
fact that the client was receiving unhoisted events. The test now also
tests the client state to detect this.

poljar committed:
base: Don't double borrow the response in one iter rooms method.

DevinR528 committed:
matrix-sdk-base: clean up recv/iter joined post rebase

The types for account data in a sync response have changed, no longer
Option. Re word comment in hoist prev_content test.

DevinR528 committed:
sdk_base: Remove room_id as argument from all Room methods

Remove room_id paramater from some client methods. Make CreationContent
two methods of RoomBuilder. Add docs for MessageWrapper.
